File: swift/Sources/TokenTrackerCore/Settings/SettingsStore.swift
//
//  SettingsStore.swift
//  TokenTrackerCore
//
//  移植自 tokentracker/prefs.py + server.py 的 _SETTINGS_SCHEMA：
//  ~/.tokentracker/settings.json 的读写、默认值与白名单校验（未列出的键拒绝）。
//  状态栏轮询文件变化热生效，无需重启。
//

import Foundation

public struct SettingsStore {
    public static let terminalApps = ["auto", "terminal", "iterm", "wezterm", "ghostty"]

    public nonisolated(unsafe) static let defaults: [String: Any] = [
        "menubar_provider": "claude",   // 状态栏标题追加显示的平台配额；"off" = 仅今日用量
        "menubar_compact": false,       // 紧凑标题（刘海屏防挤出）
        "menubar_ring": true,           // 圆环显示配额（彩色扇形圆代替 ⚡）
        "launch_at_login": false,       // 开机自动启动（SMAppService）
        "terminal_app": "auto",         // 继续会话用哪个终端
        "unit_yi": false,               // 大数以「亿」显示
        "scan_interval": 60,            // 自动扫描/刷新节奏（秒）
        "price_sync_enabled": true,     // 每 24 小时自动同步官方公开 API 费率

        // 公开统计上报。默认关闭不可协商 —— 这是个把数据发到公网的开关。
        "publish_enabled": false,       // 扫描结束后自动上报
        "publish_endpoint": "",         // 形如 https://tt.example.com（必须 https）
        "publish_handle": "",           // 服务上的用户名
        "publish_days": 365,            // 热力图窗口天数
    ]

    public let path: String

    public init(path: String? = nil) {
        self.path = path
            ?? NSHomeDirectory() + "/.tokentracker/settings.json"
    }

    public func load() -> [String: Any] {
        guard let data = FileManager.default.contents(atPath: path),
              let obj = try? JSONSerialization.jsonObject(with: data),
              let dict = obj as? [String: Any] else { return [:] }
        return dict
    }

    /// 原子写入（tmp + rename），对齐 Python save_prefs。
    public func save(_ prefs: [String: Any]) {
        let dir = (path as NSString).deletingLastPathComponent
        try? FileManager.default.createDirectory(atPath: dir, withIntermediateDirectories: true)
        guard let data = try? JSONSerialization.data(withJSONObject: prefs) else { return }
        let tmp = path + ".tmp"
        FileManager.default.createFile(atPath: tmp, contents: data)
        _ = try? FileManager.default.replaceItemAt(URL(fileURLWithPath: path),
                                                   withItemAt: URL(fileURLWithPath: tmp))
        // replaceItemAt 对已存在目标原子替换；目标不存在时回退 rename
        if FileManager.default.fileExists(atPath: tmp) {
            try? FileManager.default.moveItem(atPath: tmp, toPath: path)
        }
    }

    // ---------------------------------------------------------- 校验 ----

    private static let providerPattern = try! NSRegularExpression(
        pattern: #"off|[a-z0-9][a-z0-9_-]{0,23}"#)
    /// 只接受 https：bearer token 走明文就是凭据泄漏。
    private static let endpointPattern = try! NSRegularExpression(
        pattern: #"https://[a-z0-9.-]{3,64}(:[0-9]{2,5})?(/[A-Za-z0-9._~/-]{0,64})?"#)
    private static let handlePattern = try! NSRegularExpression(
        pattern: #"[a-z0-9][a-z0-9-]{1,30}"#)

    private static func fullMatch(_ regex: NSRegularExpression, _ v: String) -> Bool {
        let range = NSRange(v.startIndex..., in: v)
        return regex.firstMatch(in: v, range: range)?.range == range
    }

    public static func isValid(key: String, value: Any) -> Bool {
        switch key {
        case "menubar_provider":
            guard let v = value as? String else { return false }
            let range = NSRange(v.startIndex..., in: v)
            return providerPattern.firstMatch(in: v, range: range)?.range == range
        case "menubar_compact", "menubar_ring", "launch_at_login", "unit_yi", "publish_enabled",
             "price_sync_enabled":
            return (value as? NSNumber).map { CFGetTypeID($0) == CFBooleanGetTypeID() } ?? false
        case "publish_endpoint":
            guard let v = value as? String else { return false }
            return v.isEmpty || (v.count <= 200 && fullMatch(endpointPattern, v))
        case "publish_handle":
            guard let v = value as? String else { return false }
            return v.isEmpty || fullMatch(handlePattern, v)
        case "publish_days":
            return [90, 365, 730].contains((value as? NSNumber)?.intValue ?? 0)
        case "terminal_app":
            return (value as? String).map { terminalApps.contains($0) } ?? false
        case "scan_interval":
            return [30, 60, 300, 600].contains((value as? NSNumber)?.intValue ?? 0)
        default:
            return false
        }
    }

    /// 对齐 _settings_payload：默认值 + 通过白名单校验的已存值。
    public func effective() -> [String: Any] {
        var settings = SettingsStore.defaults
        for (key, value) in load() where SettingsStore.isValid(key: key, value: value) {
            settings[key] = value
        }
        return settings
    }

    /// 写入单个键（先校验，与现有偏好合并后原子保存）。
    @discardableResult
    public func set(key: String, value: Any) -> Bool {
        guard SettingsStore.isValid(key: key, value: value) else { return false }
        var prefs = load()
        prefs[key] = value
        save(prefs)
        return true
    }

    /// 一次校验并原子保存多个设置，避免表单保存到一半时被 5 秒轮询读到。
    @discardableResult
    public func set(values: [String: Any]) -> Bool {
        guard values.allSatisfy({ SettingsStore.isValid(key: $0.key, value: $0.value) })
        else { return false }
        var prefs = load()
        for (key, value) in values { prefs[key] = value }
        save(prefs)
        return true
    }

    // 便捷读取
    public func effectiveString(_ key: String) -> String? { effective()[key] as? String }
    public func effectiveBool(_ key: String) -> Bool {
        (effective()[key] as? NSNumber)?.boolValue ?? false
    }
    public func effectiveInt(_ key: String) -> Int? { (effective()[key] as? NSNumber)?.intValue }
}

File: windows/src/TokenTracker.Core/Settings/SettingsStore.cs
using System.Text.RegularExpressions;
using TokenTracker.Core.Json;
using TokenTracker.Core.Platform;

namespace TokenTracker.Core.Settings;

/// <summary>
/// ~/.tokentracker/settings.json 的读写、默认值与白名单校验（移植 SettingsStore.swift；未列出的键拒绝）。
/// 与 macOS / Python 版共用同一文件与键。
/// </summary>
public sealed class SettingsStore(string? path = null)
{
    public static readonly string[] TerminalApps = ["auto", "terminal", "iterm", "wezterm", "ghostty"];
    public static readonly long[] ScanIntervals = [30, 60, 300, 600];

    public static IReadOnlyDictionary<string, object?> Defaults { get; } = new Dictionary<string, object?>
    {
        ["menubar_provider"] = "claude", // 托盘追加显示的平台配额；"off" = 仅今日用量
        ["menubar_compact"] = false,
        ["menubar_ring"] = true, // 圆环显示配额
        ["launch_at_login"] = false,
        ["terminal_app"] = "auto",
        ["unit_yi"] = false, // 大数以「亿」显示
        ["scan_interval"] = 60L, // 自动扫描/刷新节奏（秒）
        ["publish_enabled"] = false,
        ["publish_endpoint"] = "",
        ["publish_handle"] = "",
        ["publish_days"] = 365L,
    };

    static readonly Regex ProviderPattern = new(@"^(?:off|[a-z0-9][a-z0-9_-]{0,23})$");
    static readonly Regex EndpointPattern = new(@"^https://[a-z0-9.-]{3,64}(:[0-9]{2,5})?(/[A-Za-z0-9._~/-]{0,64})?$");
    static readonly Regex HandlePattern = new(@"^[a-z0-9][a-z0-9-]{1,30}$");

    public string Path { get; } = path ?? System.IO.Path.Combine(WinPaths.DataDir, "settings.json");

    public Dictionary<string, object?> Load() => PyJson.ReadObjectFile(Path) ?? new Dictionary<string, object?>();

    public bool Save(Dictionary<string, object?> prefs) => SharedFile.AtomicWriteText(Path, PyJson.Serialize(prefs));

    public static bool IsValid(string key, object? value) => key switch
    {
        "menubar_provider" => value is string v && ProviderPattern.IsMatch(v),
        "menubar_compact" or "menubar_ring" or "launch_at_login" or "unit_yi" or "publish_enabled" => value is bool,
        "publish_endpoint" => value is string v && (v.Length == 0 || (v.Length <= 200 && EndpointPattern.IsMatch(v))),
        "publish_handle" => value is string v && (v.Length == 0 || HandlePattern.IsMatch(v)),
        "publish_days" => PyJson.AsNumber(value) is 90 or 365 or 730,
        "terminal_app" => value is string v && TerminalApps.Contains(v),
        "scan_interval" => value is long or int or double && ScanIntervals.Contains((long)PyJson.AsNumber(value)!.Value),
        _ => false,
    };

    /// <summary>默认值 + 通过白名单校验的已存值。</summary>
    public Dictionary<string, object?> Effective()
    {
        var settings = new Dictionary<string, object?>(Defaults);
        foreach (var (key, value) in Load())
            if (IsValid(key, value)) settings[key] = value;
        return settings;
    }

    /// <summary>一次校验并原子保存多个设置（与现有偏好合并）。</summary>
    public bool Set(IReadOnlyDictionary<string, object?> values)
    {
        if (!values.All(kv => IsValid(kv.Key, kv.Value))) return false;
        var prefs = Load();
        foreach (var (key, value) in values) prefs[key] = value;
        return Save(prefs);
    }

    public bool Set(string key, object? value) => Set(new Dictionary<string, object?> { [key] = value });

    public string String(string key) => Effective().Get(key) as string ?? "";
    public bool Bool(string key) => Effective().Get(key) is true;
    public long Int(string key) => PyJson.AsLong(Effective().Get(key)) ?? 0;

    /// <summary>文件指纹：mtime + 大小（设置被 CLI / 其他实例修改时热重载）。</summary>
    public string Fingerprint()
    {
        try
        {
            var info = new FileInfo(Path);
            return info.Exists ? $"{info.LastWriteTimeUtc.Ticks}:{info.Length}" : "missing";
        }
        catch (IOException)
        {
            return "error";
        }
    }
}
